### ISSUE TITLE
cannot resolve lxml.etree

### ISSUE THREAD

**Issue by johnniemorrow** [astral-sh/ty#487]
### Summary

Hi ty-team,

I was trying to fix up some unresolved-attribute errors reported by ty related to lxml.etree in [lxml](https://pypi.org/project/lxml/).

Reducing it down to just an import statement, ty doesn't seem to be able to see lxml.etree:

```bash
$ cat use_lxml.py 
import lxml.etree

```

```bash
$ ty check use_lxml.py 
WARN ty is pre-release software and not ready for production use. Expect to encounter bugs, missing features, and fatal errors.
error[unresolved-import]: Cannot resolve imported module `lxml.etree`
 --> use_lxml.py:1:8
  |
1 | import lxml.etree
  |        ^^^^^^^^^^
  |
info: make sure your Python environment is properly configured: https://github.com/astral-sh/ty/blob/main/docs/README.md#python-environment
info: rule `unresolved-import` is enabled by default

Found 1 diagnostic

```

**Comment by sharkdp** [comment]
Thank you for reporting this.

`lxml.etree` is a C extension module. I wonder if we could detect this and show a better error?

You probably want to fix this by installing `types-lxml`?

**Comment by AlexWaygood** [comment]
> `lxml.etree` is a C extension module. I wonder if we could detect this and show a better error?

Pyright does a fancy thing in its module resolver to distinguish between modules that are definitely unresolvable (no python or C-extension modules are available) and modules which might be resolvable but which it cannot inspect, because it's a C extension or similar and there are no stubs installed.

Mypy does not have that feature, but it does have a hardcoded list of untyped packages for which it knows high-quality stubs are available, and issues a nice diagnostic suggesting to install the relevant stubs package if it issues an `unresolved-import` diagnostic on one of those packages.

We don't have either feature right now; we could consider adding one or both. (The mypy one is easier to implement!)

**Comment by johnniemorrow** [comment]
> Thank you for reporting this.
> 
> `lxml.etree` is a C extension module. I wonder if we could detect this and show a better error?
> 
> You probably want to fix this by installing `types-lxml`?

`uv add types-lxml --dev` does the trick - thanks @sharkdp!

**Comment by AlexWaygood** [comment]
#715 includes some discussion on exactly which extensions should be recognized when resolving modules to C extensions, and some other details of the implementation.

**Comment by carljm** [comment]
#1613 raises the same scenario, and also suggests that we might be able to automatically introspect the `.so` and generate some implicit default stubs automatically.

**Comment by dabelknap** [comment]
I'm also seeing this issue when using pyo3 and maturin.
```
$ maturin new -b pyo3 mylib
$ cd mylib
```
By default, the contents of `src/lib.rs` are:
```Rust
use pyo3::prelude::*;

/// A Python module implemented in Rust.
#[pymodule]
mod mylib {
    use pyo3::prelude::*;

    /// Formats the sum of two numbers as string.
    #[pyfunction]
    fn sum_as_string(a: usize, b: usize) -> PyResult<String> {
        Ok((a + b).to_string())
    }
}
```
```
$ uv sync  # Setup a venv and build/install the pyo3 library (using a Homebrew install of Python 3.14 on macOS)
$ source .venv/bin/activate
```
Create a test script
```Python
# main.py
from mylib import sum_as_string
print(sum_as_string(2, 3))
```
Running the script should be successful:
```
(mylib) $ uv run main.py
5
(mylib) $
```
The import is clearly valid, but `ty` emits an error:
```
(mylib) $ ty check
error[unresolved-attribute]: Module `mylib` has no member `sum_as_string`
 --> main.py:3:7
  |
1 | import mylib as ml
2 |
3 | print(ml.sum_as_string(2, 3))
  |       ^^^^^^^^^^^^^^^^
  |
info: rule `unresolved-attribute` is enabled by default

Found 1 diagnostic
```

**Comment by konstin** [comment]
While long term pyo3 and maturin should generate type stubs (https://github.com/PyO3/maturin/pull/2940), attributes on native modules are something that imho should be ignored by default instead of erroring: The problem is a missing `.pyi` file, rather than not fulfilling a known type contract.

**Comment by MichaReiser** [comment]
>  attributes on native modules are something that imho should be ignored by default instead of erroring

This is also what pyrefly does

> When a .pyc file is encountered and no source/stub files are available, Pyrefly automatically treats module as typing.Any. This behavior ensures that compiled Python files without available source code do not cause import errors and are handled permissivel ([source](https://pyrefly.org/en/docs/configuration/#ignore-missing-imports))

**Comment by AlexWaygood** [comment]
I strongly feel that if we see that a C extension is installed without stubs, we should emit some kind of diagnostic warning the user that we cannot see inside that module and that all attribute access and imports from that module will resolve to `Unknown`. I also think it should be a separate error code to `unresolved-import` with a clear error message that tells you that the resolution is either to install stubs for the module (ideally we would tell you which stubs package to install), or to list the module in the setting added in https://github.com/astral-sh/ruff/pull/22800. I'm fine with it being a warning-level diagnostic if we think that's appropriate, but from a type-checking point of view I think we should very rarely silently fallback to `Unknown` without telling the user unless they've explicitly selected configuration to opt into that.

Now, obviously we'll have some LSP users who aren't interested in type-checking at all, and if we create an alternative "profile" for those users to opt into then this rule probably shouldn't be enabled as part of that profile. But possibly `unresolved-import` shouldn't be either?

**Comment by konstin** [comment]
Splitting `unresolved-import` into a "this import doesn't exist in the source (unless you're doing some evil `setattr` things)" and "we can't inspect the source (it's a native module)" would mitigate a lot of this.

**Comment by AlexWaygood** [comment]
> Splitting `unresolved-import` into a "this import doesn't exist in the source (unless you're doing some evil `setattr` things)" and "we can't inspect the source (it's a native module)" would mitigate a lot of this.

yeah, that's what this issue is about! :-)

**Comment by Danferno** [comment]
> > Thank you for reporting this.
> > `lxml.etree` is a C extension module. I wonder if we could detect this and show a better error?
> > You probably want to fix this by installing `types-lxml`?
> 
> `uv add types-lxml --dev` does the trick - thanks [@sharkdp](https://github.com/sharkdp)!

This also fixes following error

```
[Error - 10:16:51 AM] Request textDocument/completion failed.
  Message: request handler panicked at crates\ty_module_resolver\src\module.rs:177:35:
No root found for path 'c:\virtualenvs\datatog-venv\Lib\site-packages\lxml'. Known roots: FileRoots(
```

Posting it here because searching for it on google did not return any results and it caused the entire ty server to crash in my case

**Comment by sharkdp** [comment]
> This also fixes following error
> 
> ```
> [Error - 10:16:51 AM] Request textDocument/completion failed.
>   Message: request handler panicked at crates\ty_module_resolver\src\module.rs:177:35:
> No root found for path 'c:\virtualenvs\datatog-venv\Lib\site-packages\lxml'. Known roots: FileRoots(
> ```
> 
> Posting it here because searching for it on google did not return any results and it caused the entire ty server to crash in my case

That really sounds like a separate problem, even if it's fixed by installing the stubs. It would be great if you could open a separate issue.

**Comment by reactive-firewall** [comment]
> Splitting `unresolved-import` into a "this import doesn't exist in the source (unless you're doing some evil `setattr` things)" and "we can't inspect the source (it's a native module)" would mitigate a lot of this.

@carljm 

> [#1613](https://github.com/astral-sh/ty/issues/1613) raises the same scenario, and also suggests that we might be able to automatically introspect the `.so` and generate some implicit default stubs automatically.

Consider something like this:

```console
>>> import ctypes
... import inspect
... import os
... import platform
... 
... # Mapping of C types to Python types
... C_TO_PY_TYPE_MAP = {
...     ctypes.c_int: int,
...     ctypes.c_float: float,
...     ctypes.c_double: float,
...     ctypes.c_char_p: str,
...     ctypes.c_void_p: object,
...     # Add more mappings as needed
... }
... 
... def get_python_type_from_symbol(shared_lib_path, symbol_name):
...     # Load the shared library
...     try:
...         lib = ctypes.CDLL(shared_lib_path)
...     except Exception as e:
...         return f"Error loading library: {str(e)}"
... 
...     # Check if the symbol exists
...     try:
...         func = getattr(lib, symbol_name)
...     except AttributeError:
...         return f"Symbol '{symbol_name}' not found in the library"
... 
...     # Get the function signature
...     if not callable(func):
...         return f"Symbol '{symbol_name}' is not callable"
... 
...     # Attempt to inspect the function's argument types
...     arg_types = []
...     for arg in inspect.getfullargspec(func).args:
...         c_type = ctypes.c_void_p  # Default type
...         if arg in C_TO_PY_TYPE_MAP:
...             c_type = C_TO_PY_TYPE_MAP[arg]
...         arg_types.append(c_type)
... 
...     # Assuming the function returns a single value
...     return_type = func.restype
... 
...     # Map C return type to Python type
...     python_return_type = C_TO_PY_TYPE_MAP.get(return_type, object)
... 
...     return python_return_type
...     
>>> # example usage
>>> target_lib_path = "/usr/local/homebrew/Cellar/python@3.14/3.14.3_1/Frameworks/Python.framework/Versions/3.14/lib/libpython3.14.dylib"  # e.g., path to the python lib 3.14 installed via homebrew
>>> target_symbol_name = 'PyLong_FromLong'  # e.g., native symbol
>>> result = get_python_type_from_symbol(target_lib_path, target_symbol_name)
>>> print(result)
<class 'int'>
>>> exit()
```

I'm not familiar with the `ty` codebase but could something like this be done for `.so`/`.dylib` stubs to provide _some_ introspection of c-types? My rational is that in the case of C-lang FFI (i.e. foreign from python's view) there is a good chance that there is a type (even if not a python type) because C-Lang is strong typed itself.

Obviously as already pointed out by @konstin,

> we should emit some kind of diagnostic warning the user that we cannot see inside that module and that all [other] attribute access and imports from that module will resolve to `Unknown`.

_Sidebar: Sorry if this is the wrong GHI for this, (the online docs pointed me here)_

:bow: hope this helps!

**Comment by phineas-pta** [comment]
the only type checker that works for me so far is `zuban` which is also written in rust, maybe the team can look at it for inspiration

for my use case, i have a small project depending on 2 packages `gmpy2` and `pyicu` (their source code are almost entirely C / C++)

i tried `mypy`, `pyright`, `pyrefly`, `ty` but they all give errors

(i'm aware that there is `gmpy2-stubs` but i find nothing for `pyicu`)

**Comment by carljm** [comment]
@phineas-pta This would be more useful with some specific sample code that you expect to not error. Otherwise it's not actionable.

**Comment by phineas-pta** [comment]
sorry for the low quality comment, you can disregard my earlier comment because my setup was messed up

i re-install `ty` and `zuban` to make proper comparison and now they both give error

`ty`:
<img src="https://github.com/user-attachments/assets/ae5a5626-aa3a-46ba-9b19-218feb338b6b" />

`zuban` also gives error but it somehow manages to provide docs (and auto-completion in subsequent use):
<img src="https://github.com/user-attachments/assets/d3a96e1b-cf32-4438-b20d-2697c12c68c0" />

**Comment by carljm** [comment]
No idea how zuban is simultaneously concluding that the attribute doesn't exist and giving its docstring etc, but I don't think that is something we'd want to emulate; if we had enough confidence about what the attribute refers to to show a docstring for it, we likely wouldn't want to say that it doesn't exist.

There's really no way for a Python type checker to provide good types for a library like gmpy2 without the typing stubs. This issue is just about providing better diagnostics clarifying _why_ we can't provide types. The other alternative is to silence the errors by using [replace-imports-with-any](https://docs.astral.sh/ty/reference/configuration/#replace-imports-with-any).

**Comment by reactive-firewall** [comment]
@carljim (and @phineas-pta)
> No idea how zuban is simultaneously concluding that the attribute doesn't exist and giving its docstring etc, but I don't think that is something we'd want to emulate; if we had enough confidence about what the attribute refers to to show a docstring for it, we likely wouldn't want to say that it doesn't exist.

The docs are coming from pythons's C/C++ API for python (e.g., `#import <Python.h>` in `gmpy2`'s case, see [their fascinating /src/pythoncapi_compat.h](https://github.com/gmpy2/gmpy2/blob/08ba10ec43ded9e7663f012e9cd112e44eb66c94/src/pythoncapi_compat.h) for how they use it to export types to python and generate python docstrings among other more usual things) specifically they are adding the Python docs via the Python module scope (remember every 'source file' in Python can be a module) as docstrings in the C source [around line 27 in  gmpy2_mpz.c](https://github.com/gmpy2/gmpy2/blob/08ba10ec43ded9e7663f012e9cd112e44eb66c94/src/gmpy2_mpz.c#L27-L43)

The error does not seem the same as what `ty` is doing, as `ty` "fails to resolve import" (and without being familiar with `ty`'s source may not be accessing it in the same way (e.g., eroniously as an attribute of the module))

Hope this helps :bow:

---

<details><summary>TL;DR: Side-bar on what is going on when importing types vs module attributes</summary>

## Short
- The class/type object itself is not an “attribute”; the module has a named attribute (a binding) that references the type object. When you do `import foo` and use `foo.Bar`, `Bar` is a module attribute whose *value* is a type object (a class). <citation src="1"></citation>

## Why
- Modules are objects with attributes stored in the module’s namespace (`module.__dict__`). Any value defined or assigned in a module — functions, classes, instances, type objects — becomes a named attribute on that module object; that attribute points to the type object, which is a separate object and not intrinsically an “attribute.” <citation src="1"></citation>  
- The language reference’s data model and the standard library `types` documentation describe type/class objects as values you can bind, pass around, and import; the `types` module documents named interpreter types and helpers for dynamic type-related operations. <citation src="1,4"></citation>

## Relevant official docs
- Data model — Language Reference (modules, types, attributes): https://docs.python.org/3/reference/datamodel.html
- Modules — Tutorial / Library reference (module objects and importing): https://docs.python.org/3/tutorial/modules.html and https://docs.python.org/3/reference/import.html
- types — Library reference (named interpreter types, dynamic type creation): https://docs.python.org/3/library/types.html

---

*Side-bar generated with assistance from an AI and manually reviewed for accuracy.*

</details>

**Comment by carljm** [comment]
@reactive-firewall Yes, I know how the docstring is available at runtime; that doesn't clarify how a Python static analysis would be finding it, which is the only question in this thread. It seems from the linked zuban issue that that's not something zuban is doing either; I suspect @phineas-pta actually still has pylance (the default vscode Python LSP server) running alongside zuban, and pylance bundles stubs for gmpy2 (I know it bundles a number of third-party stubs packages by default), so that docstring help isn't coming from zuban either, it's coming from the bundled pylance stubs.

> The error does not seem the same as what `ty` is doing, as `ty` "fails to resolve import" (and without being familiar with `ty`'s source may not be accessing it in the same way (e.g., eroniously as an attribute of the module))

I don't know what you mean here; both zuban and ty are emitting exactly the same diagnostic here, and for the same reason -- the C-defined members of the `gmpy2` module are not visible to a Python-level static analysis.

**Comment by reactive-firewall** [comment]
> I don't know what you mean here; both zuban and ty are emitting exactly the same diagnostic here, and for the same reason -- the C-defined members of the `gmpy2` module are not visible to a Python-level static analysis.

I mean that in the `Module "gmpy2" has no attribute "mpz" zuban (attr-defined)` case the error is correct; from the sourcecode it is clear that `mzp` is a C/C++ implementation of a _Python type_ and that a _Python type_ is indeed _not_ an attribute of the module `gmpy2` and thus
> Module "gmpy2" has no attribute "mpz"
is technically correct (when assuming `gmpy2` has been imported (and thus can provide the py docstrings)). :shrug:

I was speculating about the odd difference from `ty` that talks about a "member" (which is _undefined_ in the Python language) may not be the same _runtime_ error as seen by Python (nor zuban), however it would make sense from the C/C++ Lang perspective to call it a member. I just wasn't sure myself about `ty`'s root cause.

Sorry for any confusion this caused.

**Comment by beatreichenbach** [comment]
Just adding that OpenEXR also hits this.
It ships only OpenEXR.cpython-313-x86_64-linux-gnu.so in site-packages with no .pyi/py.typed, so import OpenEXR is reported as unresolved-import even though it imports fine at runtime.

I will request the stubs from them separately.

**Comment by MichaReiser** [comment]
@beatreichenbach You can use https://docs.astral.sh/ty/reference/configuration/#allowed-unresolved-imports to silence the unresolved imports globally instead of having to silence each of them. Obviously, stubs would be better, because you get proper types. Another alternative is to define the stubs yourself and add them to your search path.
